### PR TITLE
#1433 toolbar button macro

### PR DESCRIPTION
**Pull request by tobibeer**
Regarding #1433:

* added toolbar-button macro
* used it in appropriate templates
* fixed templates that didn't need to specify style-classes
* fixed recently added button styles for text-buttons (compare tiddler-info-tools)
    * addendum: needs more fixing => new issue / PR

demo: http://1433.tiddlyspot.com

### CODE REVIEW

File: themes/tiddlywiki/vanilla/base.tid [hunk 1]
```diff
@@ -333,16 +333,17 @@ button svg, button img {
...
 	margin: 1px;
+	line-height: 1em;
 	background: none;
 	border: 1px solid <<colour tiddler-controls-foreground>>;
 	border-radius: 0.25em;
 }
 
 html body.tc-body .tc-btn-boxed svg {
-	font-size: 1.6666em;
+	margin-top:-0.1em;
```
Comment (pmario) [inline themes/tiddlywiki/vanilla/base.tid]: hmm, why did you need such strange CSS rules. ... Are there some new DOM elements?

File: core/ui/TiddlerInfo/Tools.tid [hunk 2]
```diff
@@ -4,22 +4,13 @@ caption: {{$:/language/TiddlerInfo/Tools/Caption}}
 
 \define lingo-base() $:/language/TiddlerInfo/
 \define config-title()
-$:/config/ViewToolbarButtons/Visibility/$(listItem)$
+$:/config/ViewToolbarButtons/Visibility/$(currentTiddler)$
 \end
-<$set name="tv-config-toolbar-icons" value="yes">
 
-<$set name="tv-config-toolbar-text" value="yes">
+<$list filter="[all[shadows+tiddlers]tag[$:/tags/ViewToolbar]!has[draft.of]]">
```
Comment (Jermolene) [inline core/ui/TiddlerInfo/Tools.tid]: It looks like you're overwriting the current tiddler with this list widget. I think that will cause problems with toolbar button handlers that reference the current tiddler, won't it?

File: themes/tiddlywiki/vanilla/base.tid [hunk 3]
```diff
@@ -333,16 +333,17 @@ button svg, button img {
 }
 
 .tc-btn-boxed {
-	font-size: 0.6em;
```
Comment (Jermolene) [inline themes/tiddlywiki/vanilla/base.tid]: What's the purpose of all the CSS changes?

Comment (Jermolene) [inline themes/tiddlywiki/vanilla/base.tid]: The worry being that they don't look backwards compatible.

File: themes/tiddlywiki/vanilla/base.tid (hunk 1 above)
Comment (tobibeer) [inline themes/tiddlywiki/vanilla/base.tid]: Which one do you mean? The deleted one or the added one?

A negative margin is nothing strange. The element wasn't centered and it wasn't asking for more padding. But yes, a font-size of 1.6666em does indeed lok somewhat strange.

The deleted one I introduced first in order to fix the wrong button sizes that were there beforehand ...when the buttons were styled as if they wouldn't come with text, too, and so the text suddenly was only 50% of what it should have been.

So, I fixed the button size (by reducing it) ...only increasing the svg.

File: core/ui/TiddlerInfo/Tools.tid (hunk 2 above)
Comment (tobibeer) [inline core/ui/TiddlerInfo/Tools.tid]: Ah, you're right, that was haphazard. Will revert.

File: themes/tiddlywiki/vanilla/base.tid (hunk 3 above)
Comment (tobibeer) [inline themes/tiddlywiki/vanilla/base.tid]: This css fixes only the new button styles as their styles were basically wrong, breaking the text size. Take the current TiddlyWiki, try to apply the new button styles while looking at the button text... and you'll see miniature text. That is, those buttons styles where designed without ever looking at a "button + text" combination. So, now both works.

**Comment by tobibeer** [comment]
back to using `<<listItem>>` (...mhhh, need to be more thorough, sorry)
updated: http://1433.tiddlyspot.com

File: themes/tiddlywiki/vanilla/base.tid (hunk 3 above)
Comment (tobibeer) [inline themes/tiddlywiki/vanilla/base.tid]: @Jermolene, if you want, I can take this out and make a new ticket for it.

Comment (Jermolene) [inline themes/tiddlywiki/vanilla/base.tid]: I think it makes sense to fix at the same time; it's just that there's a lot of changes here for me to review properly.

Comment (Jermolene) [inline themes/tiddlywiki/vanilla/base.tid]: The sizing discrepancy had bothered me before.

Comment (tobibeer) [inline themes/tiddlywiki/vanilla/base.tid]: Ah, now I see. I am now using the actual button styles in the Tools panels both tiddler and sidebar... and doing that I saw that the "icon + text" variant didn't work out at all... and that it would only be ok if the buttons had no style at all, which seemed wrong as well, because after all, there are button styles... and so they should apply consistently.

Comment (tobibeer) [inline themes/tiddlywiki/vanilla/base.tid]: Ok, you are right, that needs more fixing. I will remove the modifications for Vanilla/base from this PR.

**Comment by Jermolene** [comment]
I think my big concern is that what we've ended up doing is making a macro that basically parameterises the transclusion of the toolbar button tiddler to allow the button parameters to be more easily set.

The downside is that it introduces a macro where there was none before, and there are significant performance implications. A macro will generally need to be reparsed and rendered during each refresh.

Anyhow, it feels as if a better solution would be to introduce a more general mechanism for parameterised transclusion, perhaps additional attributes on the transclude widget.

**Comment by tobibeer** [comment]
Apart from the core, this macro simplifies using buttons for everyone, e.g. in some custom (list-)ui.

Yes, there might be a performance impact on using macro to render buttons. I have not performed any tests. Will see if I can't open some 1000 tiddlers to verify what difference that makes.

**Comment by pmario** [comment]
You can set the const PERFORMANCE_INSTRUMENTATION in startup.js to true and get some info see: https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/startup/startup.js#L21

**Comment by Jermolene** [comment]
@tobibeer maybe it does make sense to break out the CSS fixes into a separate pull request, it would be good to land those I think?

**Comment by tobibeer** [comment]
Quick question: How can I invoke **addToStory** in the browser console to open a tiddler quick-n-dirty?

**Comment by Jermolene** [comment]
> Quick question: How can I invoke addToStory in the browser console to open a tiddler quick-n-dirty?

There's not a good way to do that at the moment.

I started refactoring the story manipulation code into a new Story object:

https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/story.js

It's not particularly convenient to call, though, because you need instantiate the object and then invoke a method. For example:

https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/wiki.js#L1184-L1187

You'll see that right now the code in story.js duplicates some code in navigator.js. I'd forgotten that I'd left that in limbo, it needs fixing quickly - do pick it up if you fancy a straightforward pull request after all your hard work on the others!

**Comment by Jermolene** [comment]
I meant to add that to deal with your original request we should instantiate a global `$tw.story` object; then it would be relatively easy to call `$tw.story.navigateTiddler("my tiddler")`.

**Comment by tobibeer** [comment]
Will do the performance testing in a tiddler for now via button.

**Comment by Jermolene** [comment]
By the way, I'm guessing you know the way to display a bunch of tiddlers that overwrites the current story:

```
$tw.wiki.addTiddler({title:"$:/StoryList",list:["one","two","three"]})
```

**Comment by tobibeer** [comment]
> maybe it does make sense to break out the CSS fixes into a separate pull request, it would be good to land those I think?

Needs more fiddling, would you agree the tools tabs both in the tiddler-info as well as sidebar should show the globally set button styles rather than browser defaults? Otherwise, the problem only occurs when...

1. a user changes the button style to one of the new ones
2. a user tries to use such a button within a tiddler with icon and text

**Comment by tobibeer** [comment]
A quick and dirty performance test seems to show no difference in opening 666 non-existing tiddlers at once and having all toolbar buttons enabled... both take 31 seconds on my machine.

**Comment by tobibeer** [comment]
considering the above performance results and the reduction in code (complexity), perhaps this macro is worth being part of the core, after all... even if just a minor refactoring

**Comment by Jermolene** [comment]
Apologies for the delay @tobibeer 

I'm afraid I'm not convinced by this change. The macro `$:/core/macros/toolbar-button` doesn't really do much other than gather some variable declarations together before performing a transclusion.

Maybe it would be better to extend the transclude widget to allow $attributes to be used to specify variables to applied to the transclusion. 

So this:

```
<$set name="tv-config-toolbar-icons" value="yes">

<$set name="tv-config-toolbar-text" value="yes">

<$set name="tv-config-toolbar-class" value="">

{{$:/core/ui/Buttons/tag-manager}}

</$set>

</$set>

</$set>
```

Could be replaced with:

```
<$transclude tiddler="$:/core/ui/Buttons/tag-manager" $tv-config-toolbar-icons="yes" $tv-config-toolbar-text="yes" tv-config-toolbar-class=""/>
```

**Comment by tobibeer** [comment]
> I'm afraid I'm not convinced by this change.

That is ok.

> Maybe it would be better to extend the transclude widget to allow $attributes to be used to specify variables to applied to the transclusion.

That sounds like a good shorthand approach, I like the idea very much. As indicated above, I will close this now and make this or that PR on some minor fixes in this one.
